sklearn/pipeline.py

```diff
@@ -831,3 +841,114 @@ def make_union(*transformers, **kwargs):
+        List of (name, transformer, column) tuples specifying the transformer
+        objects to be applied to subsets of the data. The columns can be
+        specified as a scalar or list (for multiple columns) of integer or
+        string values. Integers are interpreted as the positional columns,
+        strings as the keys of `X`. In case of positional integers, also
+        a slice object is accepted.
+
+    n_jobs : int, optional
+        Number of jobs to run in parallel (default 1).
+
+    transformer_weights : dict, optional
+        Multiplicative weights for features per transformer.
+        Keys are transformer names, values the weights.
+
+    Examples
+    --------
+    >>> from sklearn.pipeline import ColumnTransformer
+    >>> from sklearn.preprocessing import Normalizer
+    >>> union = ColumnTransformer(
+    ...     [("norm1", Normalizer(norm='l1'), 'subset1'),
+    ...      ("norm2", Normalizer(norm='l1'), 'subset2')])
+    >>> X = {'subset1': [[0., 1.], [2., 2.]], 'subset2': [[1., 1.], [0., 1.]]}
+    >>> union.fit_transform(X)    # doctest: +NORMALIZE_WHITESPACE
+    array([[ 0. ,  1. ,  0.5,  0.5],
+           [ 0.5,  0.5,  0. ,  1. ]])
+
+    """
+    def _iter(self, X=None, skip_none=True):
+        """Generate (name, trans, column, weight) tuples
+        """
+        get_weight = (self.transformer_weights or {}).get
+        return ((name, trans, _getitem(X, column), get_weight(name))
+                for name, trans, column in self.transformer_list
+                if not skip_none or trans is not None)
+
+    def _update_transformer_list(self, transformers):
+        transformers = iter(transformers)
+        self.transformer_list[:] = [
+            (name, None if old is None else next(transformers), column)
+            for name, old, column in self.transformer_list
+        ]
+
+
+def _getitem(X, column):
+    """
+    Get feature column from input data (array, dataframe, dict)
+
+    """
+    if X is None:
+        return X
+
+    # check whether we have string column names or integers
+    if (isinstance(column, int) or isinstance(column, slice)
+            or (isinstance(column, list)
+                and all(isinstance(col, int) for col in column))):
+        column_names = False
+    elif (isinstance(column, six.string_types)
+          or (isinstance(column, list)
+              and all(isinstance(col, six.string_types) for col in column))):
+        column_names = True
+    else:
+        raise ValueError("no valid 'column' type")
+
+    if column_names:
+        if hasattr(X, 'loc'):
+            # pandas dataframes
+            if not isinstance(column, list):
+                column = [column]
+            return X.loc[:, column]
+        else:
+            # dicts or numpy recarrays
+            if not isinstance(column, list):
+                return _ensure_2d(X[column])
+            else:
+                return np.hstack([_ensure_2d(X[col]) for col in column])
```

Comment:
Good catch. Yes, this will currently not work at all.

We can check for this and use `sparse.hstack` for sparse arrays (and also update the ensure_2d to work with it). But then you would also need to check that each column of the dict is a sparse one, and not a mixture, because then you would need to convert the sparse ones to dense, ...

Or we can say we don't support gettting multiple columns from dictionaries to avoid this. Ideas?